File: prototype2/scripts/landmark_capture.js
// Injected only into the ephemeral VWorld page; never persist the API key.
const pilot=window.__landmark={ready:true};
const arr=v=>[v.x,v.y,v.z];
const add=(a,b,s)=>C.Cartesian3.add(a,C.Cartesian3.multiplyByScalar(b,s,new C.Cartesian3()),new C.Cartesian3());
pilot.height=()=>viewer.scene.globe.getHeight(C.Cartographic.fromDegrees(SPEC.lon,SPEC.lat));
pilot.prepare=(height,ortho)=>{
  if(!Number.isFinite(height))throw Error('Target terrain is missing');
  const target=C.Cartesian3.fromDegrees(SPEC.lon,SPEC.lat,height+80);
  viewer.camera.lookAt(target,new C.HeadingPitchRange(C.Math.toRadians(CAMERA.heading_deg),C.Math.toRadians(CAMERA.pitch_deg),CAMERA.range_m));
  const f=ortho?new C.OrthographicFrustum():new C.PerspectiveFrustum();
  f.aspectRatio=1;f.near=1;f.far=10000;
  if(ortho)f.width=800;else f.fov=2*Math.atan(800/(2*CAMERA.range_m));
  viewer.camera.frustum=f;pilot.terrain=height;
  const camera=viewer.camera;
  pilot.expected={position:arr(camera.positionWC),direction:arr(camera.directionWC),up:arr(camera.upWC)};
  pilot.bars=[-250,0,250].map(d=>{
    const p=add(add(target,camera.directionWC,d),camera.upWC,d/2);
    return [add(p,camera.rightWC,-50),add(p,camera.rightWC,50)];
  });
  document.querySelector('#state').style.display='none';viewer.scene.requestRender();
};
pilot.measure=()=>{
  const camera=viewer.camera,f=camera.frustum,m=f.projectionMatrix;
  const project=C.SceneTransforms.worldToWindowCoordinates||C.SceneTransforms.wgs84ToWindowCoordinates;
  const point=p=>{const q=project(viewer.scene,p);if(!q)throw Error('Reference projection failed');return [q.x,q.y];};
  const lines=pilot.bars.map(b=>b.map(point));
  return {viewport:[innerWidth,innerHeight],drawing_buffer:[viewer.scene.drawingBufferWidth,viewer.scene.drawingBufferHeight],
    orthographic:f instanceof C.OrthographicFrustum,width:f.width??null,matrix:Array.from({length:16},(_,i)=>m[i]),
    position:arr(camera.positionWC),direction:arr(camera.directionWC),up:arr(camera.upWC),expected:pilot.expected,
    lines,lengths:lines.map(([a,b])=>Math.hypot(a[0]-b[0],a[1]-b[1])),terrain:pilot.terrain,
    ground_anchor:point(C.Cartesian3.fromDegrees(SPEC.lon,SPEC.lat,pilot.terrain)),
    tiles_loaded:Boolean(viewer.scene.globe.tilesLoaded),cesium_version:C.VERSION};
};

File: prototype2/scripts/landmark_capture.py
"""VWorld-only 1024px/800m orthographic source gate for two landmarks."""
import argparse
import copy
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
import math
import os
from pathlib import Path
import threading
import time
import seam_lab as seam

qs=seam.api.qs
vw=seam.vw
ROOT=seam.P2/'eval/vworld/landmark_pilot'
CONFIG=seam.P2/'configs/landmark_pilot.json'


def check(samples):
    try:
        if len(samples)<4:raise ValueError('Four settled samples required')
        baseline=samples[0]['expected']
        for s in samples:
            if s['viewport']!=[1024,1024] or s['drawing_buffer']!=[1024,1024]:raise ValueError('Pixel scale changed')
            if not s['orthographic'] or abs(s['width']-800)>1e-6:raise ValueError('Not the approved orthographic frame')
            m=s['matrix']
            if len(m)!=16 or not all(math.isfinite(v) for v in m) or abs(m[15]-1)>1e-9 or abs(m[11])>1e-9:raise ValueError('Invalid projection matrix')
            if len(s['lengths'])!=3 or any(not math.isfinite(v) or abs(v-128)>1 for v in s['lengths']):raise ValueError('Scale varies with depth')
            for field,tol in [('position',.01),('direction',1e-6),('up',1e-6)]:
                if len(s[field])!=3 or not all(math.isfinite(v) for v in s[field]) or math.dist(s[field],baseline[field])>tol:raise ValueError('Camera drift')
            if len(s['ground_anchor'])!=2 or any(not math.isfinite(v) or v<0 or v>1024 for v in s['ground_anchor']):raise ValueError('Missing ground anchor')
        return {'passed':True,'reference_100m_px':samples[-1]['lengths'],'ground_anchor':samples[-1]['ground_anchor']}
    except (KeyError,TypeError,ValueError) as e:return {'passed':False,'reason':str(e)}


def page(key,scene):
    cfg=qs.read(CONFIG);spec={**cfg['scenes'][scene],'id':scene,'size':1024,'meters_per_pixel':800/1024}
    source_cfg=copy.deepcopy(vw.read(vw.CONFIG));source_cfg['camera']=cfg['camera'];source_cfg['try_orthographic']=False
    original=vw.page(key,spec,source_cfg)
    start=original.index('      const applyCamera=()=>{');end=original.index('    } catch(e){fail(e)}',start)
    return original[:start]+Path(__file__).with_suffix('.js').read_text()+original[end:]


def capture(scene):
    key=os.environ.get('VWORLD_API_KEY','').strip()
    if not key:raise ValueError('VWORLD_API_KEY is missing')
    cfg=qs.read(CONFIG)
    if cfg['size']!=1024 or cfg['width_m']!=800 or cfg['camera']!={'heading_deg':22.5,'pitch_deg':-30,'range_m':900}:raise ValueError('Approved camera changed')
    content=page(key,scene).encode()
    class Handler(BaseHTTPRequestHandler):
        def do_GET(self):
            if self.path!='/':self.send_error(404);return
            self.send_response(200);self.send_header('Content-Type','text/html; charset=utf-8');self.send_header('Cache-Control','no-store');self.end_headers();self.wfile.write(content)
        def log_message(self,*_):pass
    server=ThreadingHTTPServer(('localhost',8767),Handler)
    folder=ROOT/'captures'/seam.stamp()/scene;folder.mkdir(parents=True)
    r={'scene':scene,'spec':cfg['scenes'][scene],'camera':cfg['camera'],'size':1024,'width_m':800,
       'samples':[],'visual_approved':False,'ai_requests':0,'key_persisted':False,'source':'국토교통부 / VWorld'}
    threading.Thread(target=server.serve_forever,daemon=True).start()
    print('CAPTURE='+str(folder.relative_to(seam.P2)),flush=True)
    try:
        from playwright.sync_api import sync_playwright
        with sync_playwright() as pw:
            browser=pw.chromium.launch(headless=True)
            tab=browser.new_page(viewport={'width':1024,'height':1024},device_scale_factor=1)
            tab.goto('http://localhost:8767/',wait_until='domcontentloaded',timeout=30000)
            tab.wait_for_function('()=>window.__landmark?.ready||window.__P2?.error',timeout=90000)
            if tab.evaluate('()=>window.__P2.error'):raise ValueError('VWorld initialization failed')
            tab.wait_for_timeout(12000)
            deadline=time.monotonic()+60;heights=[]
            while time.monotonic()<deadline:
                h=tab.evaluate('()=>window.__landmark.height()')
                if isinstance(h,(int,float)) and math.isfinite(h):heights.append(h)
                else:heights=[]
                if len(heights)>=3 and max(heights[-3:])-min(heights[-3:])<.1:break
                tab.wait_for_timeout(2000)
            else:raise ValueError('Terrain height unavailable')
            tab.evaluate('h=>window.__landmark.prepare(h,false)',h)
            tab.wait_for_timeout(12000)  # Building cache only; no perspective input is saved.
            tab.evaluate('h=>window.__landmark.prepare(h,true)',h)
            stable=0
            for i in range(45):
                tab.wait_for_timeout(2000);s=tab.evaluate('()=>window.__landmark.measure()');r['samples'].append(s)
                stable=stable+1 if s['tiles_loaded'] else 0
                if (i+1)%10==0:print(scene+': waiting for orthographic source',flush=True)
                if i>=29 and stable>=3:break
            r['samples'].append(tab.evaluate('()=>window.__landmark.measure()'))
            r['validation']=check(r['samples'])
            tab.screenshot(path=str(folder/'source.png'));browser.close()
    except Exception as e:
        r['error']=vw.redact(str(e),key);r['validation']={'passed':False,'reason':'Capture failed'}
    finally:
        server.shutdown();server.server_close()
        r['files']={p.name:qs.sha(p) for p in folder.glob('*.png')}
        vw.write(folder/'capture.json',vw.secret_safe(r,key))
    print(scene+': '+str(r['validation']),flush=True)
    return folder


if __name__=='__main__':
    p=argparse.ArgumentParser(description=__doc__);p.add_argument('--scene',choices=['all','gwanghwamun','jongno-tower'],default='all');p.add_argument('--allow-external',action='store_true');a=p.parse_args()
    if not a.allow_external:raise ValueError('Explicit VWorld opt-in required')
    for scene in qs.read(CONFIG)['scenes'] if a.scene=='all' else [a.scene]:capture(scene)
